include/seqan3/io/alignment_file/output.hpp

```diff
@@ -0,0 +1,687 @@
+     *
+     * Basic exception safety.
+     *
+     * ### Example
+     *
+     * ```cpp
+     * alignment_file_output fout{"/tmp/my.sam"};
+     *
+     * std::vector<std::tuple<dna5_vector, std::string>> range
+     * {
+     *     { "ACGT"_dna5, "First" },
+     *     { "NATA"_dna5, "2nd" },
+     *     { "GATA"_dna5, "Third" }
+     * }; // a range of "records"
+     *
+     * fout = range; // will iterate over the records and write them
+     * ```
+     */
+    template <typename rng_t>
+    alignment_file_output & operator=(rng_t && range)
+        requires std::ranges::InputRange<rng_t> && tuple_like_concept<reference_t<rng_t>>
+    {
+        for (auto && record : range)
+            push_back(std::forward<decltype(record)>(record));
+    }
+
+    /*!\brief            Write a range of records (or tuples) to the file.
+     * \tparam rng_t     Type of the range, must satisfy seqan3::std::ranges::InputRange and have a reference type that
+     *                   satisfies seqan3::tuple_like_concept.
+     * \param[in] range  The range to write.
+     * \param[in] f      The file being written to.
+     *
+     * \details
+     *
+     * This operator enables alignment_file_output to be at the end of a piping operation. It just calls
+     * operator=() internally.
+     *
+     * ### Complexity
+     *
+     * Linear in the number of records.
+     *
+     * ### Exceptions
+     *
+     * Basic exception safety.
+     *
+     * ### Example
+     *
+     * \snippet test/snippet/io/alignment_file/output.cpp write_range
+     *
+     * This is especially useful in combination with file-based filters:
+     *
+     * \todo TODO Implement the snippet once seqan3::alignment_file_in is implemented
+     *
+     * ```cpp
+     * alignment_file_in{"input.sam"} | view::minimum_average_quality_filter(20)
+     *                                 | view::minimum_alignment_length_filter(50)
+     *                                 | ranges::view::take(5)
+     *                                 | alignment_file_output{"output.sam"};
+     * ```
+     */
+    template <typename rng_t>
+    friend alignment_file_output & operator|(rng_t && range, alignment_file_output & f)
+        requires std::ranges::InputRange<rng_t> && tuple_like_concept<reference_t<rng_t>>
+    {
+        f = range;
+        return f;
+    }
+
+    //!\overload
+    template <typename rng_t>
+    friend alignment_file_output operator|(rng_t && range, alignment_file_output && f)
+        requires std::ranges::InputRange<rng_t> && tuple_like_concept<reference_t<rng_t>>
+    {
+        f = range;
+        return std::move(f);
+    }
+    //!\}
+
+    //!\brief The options are public and its members can be set directly.
+    alignment_file_output_options options;
+
+    /*!\brief The file header object.
+     *
+     * ### Example
+     * \snippet test/snippet/io/alignment_file/output.cpp set_header
+     */
+    std::unique_ptr<alignment_file_header> header{nullptr};
```

Comment:
make this private and add a member function `alignment_file_header & header()` that returns a reference to the contained header and initialises the pointer if it is `nullptr`. Also use this function when passing the header to the format, thereby we hide all the pointer semantics from the user and also the format.